fix(compiler) - Compile code inside a try block

wilzbach opened the pull request:
Fixes https://github.com/storyscript/storyscript/issues/449

**- What I did**

- StoryError no longer exists directly in Story
- Passing `debug` around everywhere is superfluous with this change
- The CLI simply catches the StoryError and is itself in charge of error handling (this is also nice because the CLI class knows whether coloring should be used, sth. the error shouldn't need to know)
- On any error, the compiler now exits with a non-zero exit code (see also: https://en.wikipedia.org/wiki/Exit_status#POSIX).
- For better testing, `SyntaxError` now already supports omitting the coloring (this should be refactored as part of #458 though)
- Errors now are caught and rethrown as `StoryError`s 
- `SyntaxError` gained a `short_message` message which can be used for short one-line messages (e.g. hover error in an editor or asyncy hub and more convenient testing)
- EDIT: Catch all unknown errors in the API and expose them as StoryError's with report instructions (a start of #459)

~~Up for discussion:~~
- using a generic compiler error code for compiler errors [EDIT: no longer using a generic error]
- format of `short_message` (it could contain e.g. the error code) [EDIT: `short_message` now uses the error codes at the beginning]

Future directions:
- whether `Story.parse` should catch all exceptions and wrap them as reportable [ICE](https://en.wikipedia.org/wiki/Compilation_error)s (see also: https://github.com/storyscript/storyscript/issues/459)
- adding a CLI flag to switch between short and long/verbose/context message styles (though we should really start to introduce a per compilation config object, s.t. we don't need to forward arguments on almost every function like it was done with `debug` - see also https://github.com/storyscript/storyscript/issues/460 for the config discussion)
- `--debug` is now superfluous


**- Description for the changelog**

`StoryError` are now thrown from the public API.

wilzbach commented on storyscript/exceptions/StoryError.py:
```diff
@@ -109,3 +122,10 @@ def echo(self):
+        A short version of the error message
+        """
+        self.process()
+        return f"Error: {self.hint()}"
```
Okay, `short_message` now includes the error code, e.g.

```
    assert e.value.short_message() == 'E0007: Missing value after `=`'
```

Vesuvium commented:
Now it returns Python's traceback from the CLI for unknown errors. For known errors, it's now impossible to get the traceback because `--debug` no longer exists.

wilzbach commented:
> Now it returns Python's traceback from the CLI for unknown errors.

Well, to be fair that's also the current behavior.
Also, exactly because of this we want to catch all unknown errors anyhow. See e.g. https://github.com/storyscript/storyscript/issues/459
Though if you want I could do this directly here.

> For known errors, it's now impossible to get the traceback because --debug no longer exists.

Ah so you want basically sth. like:

```diff
diff --git a/storyscript/Cli.py b/storyscript/Cli.py
index e11d911..e5ee1c4 100644
--- a/storyscript/Cli.py
+++ b/storyscript/Cli.py
@@ -52,8 +52,11 @@ class Cli:
                 else:
                     click.echo(tree.pretty())
         except StoryError as e:
-            e.echo()
-            exit(1)
+            if debug:
+                raise e
+            else:
+                e.echo()
+                exit(1)
 
     @staticmethod
     @main.command()
@@ -83,14 +86,18 @@ class Cli:
                     msg = 'Script syntax passed!'
                     click.echo(click.style(msg, fg='green'))
         except StoryError as e:
-            e.echo()
-            exit(1)
+            if debug:
+                raise e
+            else:
+                e.echo()
+                exit(1)
 
     @staticmethod
     @main.command()
     @click.argument('path', default=os.getcwd())
+    @click.option('--debug', is_flag=True)
     @click.option('--ebnf', help=ebnf_help)
-    def lex(path, ebnf):
+    def lex(path, debug, ebnf):
         """
         Shows lexer tokens for given stories
         """
@@ -101,8 +108,11 @@ class Cli:
                 for n, token in enumerate(tokens):
                     click.echo('{} {} {}'.format(n, token.type, token.value))
         except StoryError as e:
-            e.echo()
-            exit(1)
+            if debug:
+                raise e
+            else:
+                e.echo()
+                exit(1)
 
     @staticmethod
     @main.command()
```

Vesuvium commented:
Storyscript (master) always returns a pretty printed error for syntax errors and expected compiler errors. However if the error is in the compiler and not expected, the error is not handled and the traceback is printed and that's what 449 is about.

wilzbach commented on storyscript/exceptions/StoryError.py:
```diff
@@ -101,11 +113,33 @@ def message(self):
+        url = 'https://github.com/storyscript/storyscript/issues'
+        return StoryError.unnamed_error((
+            f'Internal error occured: {str(e)}\n'
+            f'Please report at {url}'))
```
This can be improved as part of https://github.com/storyscript/storyscript/issues/459

Vesuvium commented:
Codecov was out of service for about a day

wilzbach commented:
OK. Rebased then and as I ran into a few merge conflicts I squashed this into one commit, s.t. if this stays open for longer and I will have an easier time merging things.

Vesuvium commented:
I was just about merging
